### PR TITLE
[FSDP] Option for eval in fp32/bf16

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104682

In https://github.com/pytorch/pytorch/pull/97645 and some follow up diffs, we made FSDP run in full precision in eval mode, even if mixed precision was specified.

However, this is probably not the best idea and we should provide a flag for users to have control over this a bit more. Adding an env var FSDP_FULL_PREC_IN_EVAL and defaulting it to off, users who want to run eval in fp32 can toggle this before wrapping model in FSDP:

os.environ["FSDP_FULL_PREC_IN_EVAL"] = "1"

Verified that unittests, APS workflow, TNT workloads can run eval appropriately with this change.

Differential Revision: [D47246556](https://our.internmc.facebook.com/intern/diff/D47246556/)

### CODE REVIEW

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -251,7 +252,10 @@ def _share_state_and_init_handle_attrs(
     root_state._all_fsdp_states = traversal_utils._get_fsdp_states(root_module)
     root_state._all_handles = root_state._exec_order_data.all_handles  # share reference
     root_state._device_mesh = _init_device_mesh(root_state)
-
+    import os
+    root_state._full_prec_in_eval = (
+        os.environ.get(_FSDP_FULL_PREC_IN_EVAL, "") == "1"
+    )
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:258]: We should move this to construction time. For example, we can check the new env var after this:
https://github.com/pytorch/pytorch/blob/adf1405909a3136b51be17992a8def4297764022/torch/distributed/fsdp/_init_utils.py#L393-L397

File: torch/distributed/fsdp/flat_param.py [hunk 2]
```diff
@@ -87,6 +87,9 @@
 # pre-backward each iteration.
 _FSDP_SKIP_WRITEBACK_CHECK = "FSDP_SKIP_WRITEBACK_CHECK"
 
+# Env var toggling whether when model is in .eval() mode, should we run in fp32
+# or the reduced precision.
+_FSDP_FULL_PREC_IN_EVAL = "FSDP_FULL_PREC_IN_EVAL"
```
Comment (awgu) [inline torch/distributed/fsdp/flat_param.py:92]: nit: Should the naming be `_FSDP_EVAL_IN_FULL_PREC = "FSDP_FULL_PREC_IN_EVAL"` or `_FSDP_USE_FULL_PREC_IN_EVAL = "FSDP_USE_FULL_PREC_IN_EVAL"`?

My reasoning here is that you save this to an attribute `_full_prec_in_eval: bool`, so I would read it like "if full precision in eval". I read it more naturally as "if eval in full precision" or "if use full precision in eval".

File: test/distributed/fsdp/test_fsdp_mixed_precision.py [hunk 3]
```diff
@@ -872,21 +900,27 @@ def test_full_precision_in_eval_buffers(self):
...
+                _module_classes_to_ignore=[],
             )
             model = TransformerWithSharedParams.init(
                 self.process_group,
                 FSDPInitMode.NO_FSDP if use_composable else FSDPInitMode.RECURSIVE,
                 CUDAInitMode.CUDA_BEFORE,
                 {"mixed_precision": mp_config},
             )
+            for fsdp_module in FSDP.fsdp_modules(model):
+                mp = fsdp_module.mixed_precision
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_mixed_precision.py:923]: Where is `mp` used? Why do we take the last one of the FSDP modules?

File: torch/testing/_internal/common_fsdp.py [hunk 4]
```diff
@@ -339,14 +339,17 @@ def init(
...
-                ),
+                )
+            fsdp_model = FSDP(
+                m,
                 fsdp_pg,
                 auto_wrap_policy=auto_wrap_policy,
                 **fsdp_kwargs,
             )
+            for fsdp_module in FSDP.fsdp_modules(fsdp_model):
+                mp = fsdp_module.mixed_precision
```
Comment (awgu) [inline torch/testing/_internal/common_fsdp.py:352]: Could you explain this change?

File: test/distributed/fsdp/test_fsdp_mixed_precision.py [hunk 5]
```diff
@@ -761,41 +766,54 @@ def test_eval_root_cast_inputs(self):
         """
         In a case where root module does not manage FSDP parameters,
         ensure that we don't cast forward inputs which could potentially
-        cause a dtype mismatch.
+        cause a dtype mismatch. Check that FSDP_FULL_PREC_IN_EVAL controls
+        this.
         """
 
+        reduce_dtype = torch.float16
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_mixed_precision.py:773]: nit: Curious why you named this `reduce_dtype` when we use it for `param_dtype`, `reduce_dtype`, and `buffer_dtype`? Maybe name it `low_prec_dtype`?

File: test/distributed/fsdp/test_fsdp_mixed_precision.py [hunk 6]
```diff
@@ -761,41 +766,54 @@ def test_eval_root_cast_inputs(self):
...
                 super().__init__()
                 self.a = nn.Linear(5, 5)
 
-            def forward(self, x):
-                assert x.dtype == torch.float32, f"Expected fp32, got {x.dtype}"
+            def forward(self, x, expect_full_prec_in_eval):
+                if expect_full_prec_in_eval:
+                    assert x.dtype == torch.float32, f"Expected fp32, got {x.dtype}"
+                else:
+                    assert x.dtype == reduce_dtype, f"Expected {reduce_dtype}, got {x.dtype}"
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_mixed_precision.py:784]: I guess the pain here is that t he `self` arg to `forward()` shadows the test class `self`, so it is hard to use `self.assertEquals()`.

File: test/distributed/fsdp/test_fsdp_mixed_precision.py [hunk 7]
```diff
@@ -860,9 +883,14 @@ def test_full_precision_in_eval_buffers(self):
             # model.eval() + forward pass should make the buffers in full prec again
             fsdp_model.eval()
             fsdp_model((inp, self, fsdp_model, mp_config, torch.float32))
-
+            # TODO: this test would be more robust if the buffer dtype was
+            # validated in the nn.Module pre-forward hook to ensure that the
+            # buffer computation takes place in the right precision.
             for buf in fsdp_model.buffers():
-                self.assertEqual(_BUFFER_ORIG_DTYPE, buf.dtype)
+                if full_prec_in_eval:
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_mixed_precision.py:890]: nit: You do `expected_dtype = ... if full_prec_in_eval else ...` above. You could do it here too for slight improvement in readability.

File: test/distributed/fsdp/test_fsdp_mixed_precision.py [hunk 8]
```diff
@@ -860,9 +883,14 @@ def test_full_precision_in_eval_buffers(self):
             # model.eval() + forward pass should make the buffers in full prec again
             fsdp_model.eval()
             fsdp_model((inp, self, fsdp_model, mp_config, torch.float32))
-
+            # TODO: this test would be more robust if the buffer dtype was
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_mixed_precision.py:886]: Make this a bootcamp/onboarding task? :)

**Review by awgu — COMMENTED** [review]
This change makes sense to me. I mainly want to double check (1) env var / attribute name  and (2) moving the initialization of the attribute to construction time instead of lazy initialization time.

File: torch/distributed/fsdp/flat_param.py (hunk 2 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/flat_param.py:92]: sounds good!

File: torch/testing/_internal/common_fsdp.py (hunk 4 above)
Comment (rohan-varma) [reply torch/testing/_internal/common_fsdp.py:352]: Debugging vestige, not needed

File: test/distributed/fsdp/test_fsdp_mixed_precision.py (hunk 8 above)
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_mixed_precision.py:886]: https://github.com/pytorch/pytorch/issues/104740

File: test/distributed/fsdp/test_fsdp_mixed_precision.py (hunk 3 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_mixed_precision.py:922]: Why is this resolved? :)

**Review by awgu — APPROVED** [review]
LGTM!

A minor nit is if you want, you could change the `full_prec_in_eval` in test_fsdp_mixed_precsion.py to `use_full_prec_in_eval` to match the new env var name (as a simple find & replace).

File: test/distributed/fsdp/test_fsdp_mixed_precision.py (hunk 3 above)
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_mixed_precision.py:922]: oops, I deleted this locally but somehow didn't push that change.

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 9]
```diff
@@ -606,6 +606,7 @@ def test_homogeneous_attributes(self):
         all_attr_name_and_values = [
             ("_use_orig_params", False, True),
             ("limit_all_gathers", False, True),
+            ("_use_full_prec_in_eval", False, True),
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_misc.py:609]: @awgu this is an attr that should be set uniformly for all handles, but isn't a direct ctor input. So I added the below logic to test, LMK if shoudl do something different.
